PR4

jokano12 opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

- [ ] I attest that the connector works and I verified by deploying and testing all the operations. 
- [ ] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [ ] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [ ] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [ ] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [ ] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [ ] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [ ] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [ ] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

jokano12 committed:
Update apiDefinition.swagger.json

jokano12 committed:
Add files via upload

jokano12 committed:
Delete apiDefinition.swagger.json

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -2364,4514 +3373,5821 @@
+    "CommitteeNomineeData": {
+      "type": "object",
+      "properties": {
+        "id": {
```
Please fix the description and title

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -770,65 +192,112 @@
-                  "description": "Address.",
-                  "title": "Address"
-                },
                 "type": {
```
Value is required

lordlydon replied:
Also, please add enum options here too

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -770,65 +192,112 @@
-                  "type": "boolean",
-                  "description": "Show In Directory",
-                  "title": "Show In Directory",
+                "url": {
```
This one is also required

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -770,65 +192,112 @@
+            "schema": {
+              "type": "object",
+              "properties": {
+                "type": {
```
This one is also required

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -770,65 +192,112 @@
+                    "YouTube"
                   ]
+                },
+                "url": {
```
This one is also required

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -930,93 +396,80 @@
           }
-        ]
+        ],
+        "summary": "Get Organization relationships",
```
Please change to `Get Organization's Relationships`, to match the format of the Individual's equivalent action

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -930,93 +396,80 @@
-            "default": "application/json",
-            "x-ms-visibility": "internal"
+            "schema": {
+              "$ref": "#/definitions/AddressSaveData"
```
I missed this one last time, can you add the enum options for the address type: Home, Work and Other.

jokano12 replied:
I can add the enum option to the ref "AddressSaveData", will this work? Or does this option not apply to all actions with "addresssavedata." If the latter is the case, I can remove the reference and add the schema with the enum in it.

lordlydon replied:
Yes, adding it to the ref would work

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -930,93 +396,80 @@
           "200": {
-            "description": "Success",
+            "description": "default",
             "schema": {}
```
Please check the response definition at http://abc.mpxdemo.com/api/help/Api/POST-api-v1-Individuals-idOrRecordNumber-CustomFieldsList

jokano12 replied:
When I add the schema, it causes the flow to fail and result in the following. However, I remove the schema and it works. And note that the schema that is missing is the response schema, not the body schema which is referenced in the error.
![Screenshot 2021-11-25 185937](https://user-images.githubusercontent.com/83984082/143512049-e2d31fba-6f58-43fe-b6ce-a2ccb5657c80.png)
![Screenshot 2021-11-25 185919](https://user-images.githubusercontent.com/83984082/143512050-ac251f98-91c7-427e-a86b-4913f609555a.png)

lordlydon replied:
There is an error with the swagger definition, the response is an array of `CustomFieldResultData` http://abc.mpxdemo.com/api/help/ResourceModel?modelName=CustomFieldResultData

jokano12 replied:
![Screenshot 2021-11-28 205754](https://user-images.githubusercontent.com/83984082/143801897-487cd987-b0c6-46c5-90e2-32a783518023.png)
Adding the response definition causes paconn to run an error with it. In addition, the flow doesnt work. I removed the response definition and paconn succeeded and the flow works.

lordlydon replied:
I don't get any errors adding it like this, can you check it again?
```
"responses": {
  "200": {
    "description": "default",
    "schema": {
      "type": "array",
      "items":{
        "$ref": "#/definitions/CustomFieldResultData"
      }
    }
  }
},
```

jokano12 replied:
I got the flow working, but I am still gettting the paconn error. 
![Screenshot 2021-11-30 145854](https://user-images.githubusercontent.com/83984082/144127127-4404fb3e-3eca-4fdb-bec4-ed3d039acaae.png)
You arent getting the error when run through paconn? What does your definition for customfieldresultdata look like?

lordlydon replied:
I didn't make any changes to the CustomFieldResultData definition:
```
"CustomFieldResultData": {
  "type": "object",
  "properties": {
    "status": {
      "type": "string",
      "description": "Status.",
      "title": "Status"
    },
    "message": {
      "type": "string",
      "description": "Message.",
      "title": "Message"
    },
    "customFieldName": {
      "type": "string",
      "description": "Custom Field Name.",
      "title": "Custom Field Name"
    }
  }
},
```

jokano12 replied:
My paconn still doesnt validate it. Any suggestions?

lordlydon replied:
Can you commit the changes so I can check?

jokano12 replied:
Im not sure whats different but, I checked paconn again with the file I committed to this PR and it validated this time.

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -1130,34 +559,36 @@
-        "summary": "Add Relationship to Individual",
-        "description": "Add Relationship to Individual.",
-        "operationId": "Add-relationship-to-individual",
+        "summary": "Update Custom Field Value",
```
Please change to `Update Individual's Custom Field Value` as there is a different endpoint to update values for Organizations

jokano12 committed:
Update apiDefinition.swagger.json

Made changes, waiting for verdict on response schema for Add or Update a list of Custom Fields Per Individual, on whether to keep empty, troubleshoot, or omit from PR

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -9271,7 +9284,12 @@
+            "Home",
+            "Work",
+            "Other"
+            [
```
Added wrong bracket here

jokano12 committed:
Update apiDefinition.swagger.json

jokano12 committed:
Update apiDefinition.swagger.json

lordlydon reviewed (commented):
Looks good.
@RjImpx, we should wait until the previous actions are merged into the Microsoft repository before merging this pull request.

lordlydon commented:
@jokano12, we merged the other PR and now there are some conflicts for this one. Can you fix these?

jokano12 committed:
JSON Beautifier to have same format as Microsoft-PR

jokano12 committed:
Merge branch 'dev' into dev

lordlydon commented on certified-connectors/Impexium/apiDefinition.swagger.json:
```diff
@@ -21,30 +21,66 @@
-  "paths": {
-    "/api/v1/Individuals/{ID or Record Number}/Phones": {
-      "post": {
+  "paths": 
```
Missing `{`

jokano12 committed:
Update apiDefinition.swagger.json
